OpenAPI conversion validation as a PR gate

zengin opened the pull request:
- adds OpenAPI reader as a submodule
- adds a step to MetadataParser tool to convert to OpenAPI
- adds a Github Action which will
  - apply the latest XSLT rules on the latest v1 and beta metadata
  - validate those transformed metadata is parsable

Fixes #62

zengin commented:
Tested the new validation rule with the broken XSLT and it failed. Reverted back that change and it succeeded.

zengin commented:
@baywet can you have a look at the Github Action failure for updating latest OpenAPI descriptions? Not sure why the push fails.

MIchaelMainer commented on tools/MetadataParser/Program.cs:
```diff
@@ -1,12 +1,35 @@
 ﻿using Microsoft.OData.Edm.Csdl;
 using System.Xml;
 using System;
+using Microsoft.OData.Edm;
```
nit: file should contain copyright header

baywet commented on scripts/run-metadata-validation.ps1:
```diff
@@ -0,0 +1,48 @@
+    [Parameter(Mandatory=$true)][string]$repoDirectory
+)
+
+$transformCsdlDirectory = Join-Path $repoDirectory "transforms/csdl"
```
```suggestion
$transformCsdlDirectory = Join-Path $repoDirectory "transforms" "csdl"
```

zengin replied:
Unfortunately `Join-Path` doesn't work with varargs and you need [weird tricks in piping](http://codebuckets.com/2016/08/20/powershell-join-path-with-multiple-parameters/) or multiple levels of function composition to get it working in single line.

Forward slash works cross platform.

baywet commented on scripts/run-metadata-validation.ps1:
```diff
@@ -0,0 +1,48 @@
+$betaSnapshot = Join-Path $repoDirectory "beta_metadata.xml"
+$v1Snapshot = Join-Path $repoDirectory "v1.0_metadata.xml"
+
+$metadataParserTool = Join-Path $repoDirectory "tools/MetadataParser/MetadataParser.csproj"
```
```suggestion
$metadataParserTool = Join-Path $repoDirectory "tools" "MetadataParser" "MetadataParser.csproj"
```

zengin replied:
See other comment

baywet commented on scripts/run-metadata-validation.ps1:
```diff
@@ -0,0 +1,48 @@
+& $transformScript -xslPath $xsltPath -inputPath $betaSnapshot -outputPath $transformedBeta
+
+Write-Host "Validating beta metadata after the transform..."
+& dotnet run -p $metadataParserTool $transformedBeta
```
nit: I'd move the dotnet run steps in the workflow, this way we'd have a better integration for linting, warning etc... with the UI

zengin replied:
I would like to be able to run this script as a self contained test in local dev box without needing to create a PR and then realizing that my XSLT is wrong. That is the reason I am keeping everything in a single ps1 file. I wish Actions had a way to run as is on a local dev environment without much modifications.

baywet replied:
yeah not having the ability to run workflows locally is a real pain. I pushed 18 commits to fix the workflow before squashing and force push....

you could have two scripts:
- one that calculates the paths and settings them as tasks output (this one
- one that calls the first one and dotnet run for the local debug experience

zengin replied:
I think this is a script that is short enough to be single file. And local testing instruction is as simple as:
> After you made your changes to the XSLT, run:
> `./scripts/run-metadata-validation.ps1 <local_metadata_checkout_folder>`

And everything is maintained in a single place.

baywet commented:
@zengin fixed the pipeline (I used your PR as guinea pig)!

zengin commented:
@baywet I added a quick fix to allow empty commits in OpenAPI action. Thanks for the other fixes you pushed.

I have addressed your comments. Can you please have another look?
